# Assignment: Data Wrangling

### Reading material: `tidy_data.pdf`

**Q1.** This question provides some practice cleaning variables which have common problems.
1. For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. For the Minnesota police use of for data, `./data/mn_police_use_of_force.csv`, clean the `subject_injury` variable, handling the NA's; this gives a value `Yes` when a person was injured by police, and `No` when no injury occurred. What proportion of the values are missing? Is this a concern? Cross-tabulate your cleaned `subject_injury` variable with the `force_type` variable. Are there any patterns regarding when the data are missing? 

**Q2.** Go to https://sharkattackfile.net/ and download their dataset on shark attacks (Hint: `GSAF5.xls`).

1. Open the shark attack file using Pandas. It is probably not a csv file, so `read_csv` won't work.
2. Drop any columns that do not contain data.
3. Clean the year variable. Describe the range of values you see. Filter the rows to focus on attacks since 1940. Are attacks increasing, decreasing, or remaining constant over time?
4. Clean the Age variable and make a histogram of the ages of the victims.
5. What proportion of victims are male?
6. Clean the `Type` variable so it only takes three values: Provoked and Unprovoked and Unknown. What proportion of attacks are unprovoked?

---

## *Phase 1: Solution without AI

Complete all questions in the following cells without generative AI. Then commit and push the
notebook before beginning Phase 2.

**Declaration:** I completed this version without generative AI.

**Student(s):** Dylan Wang

**Date:** 8/31/2026

In [21]:
# Your Phase 1 solution to the problem goes here.

# imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [22]:
airbnb = pd.read_csv('./data/airbnb_hw.csv')
# Q1.1:

# inspect the raw data
print(airbnb["Price"].describe(), "\n")

# convert the price column to numeric, remove the , for thousands
airbnb["Price"] = pd.to_numeric(
airbnb["Price"].str.replace(",", "", regex = False),  
errors = "coerce")

print(airbnb["Price"].describe())
print("missing values: ", airbnb["Price"].isnull().sum())


count     30478
unique      511
top         150
freq       1481
Name: Price, dtype: object 

count    30478.000000
mean       163.589737
std        197.785454
min         10.000000
25%         80.000000
50%        125.000000
75%        195.000000
max      10000.000000
Name: Price, dtype: float64
missing values:  0


The price column was stored as an object since values bigger than 999 contained commas, like 1,000. I removed the commas before force the type to be numeric. 
After the conversion, there were no missing values.

In [26]:
# Q1.2:
police_data = pd.read_csv('./data/mn_police_use_of_force.csv')

print(police_data["subject_injury"].value_counts(dropna = False))
injury = police_data["subject_injury"]

# keep yes and no, and the rest should be labeled as missing
police_data["subject_injury_cleaned"] = injury.where(
    injury.isin(["Yes", "No"]),
    "Missing"
)

print("\n", police_data["subject_injury_cleaned"].value_counts(dropna = False))

missing_percentage = (
    police_data["subject_injury_cleaned"].value_counts(dropna = False)["Missing"] / len(police_data) * 100
)

print("Percentage of missing values: ", missing_percentage, "%")
print("Percentage of non-missing values: ", 100 - missing_percentage, "%")

injury_table = pd.crosstab(
    police_data["subject_injury_cleaned"],
    police_data["force_type"]
)

display(injury_table)



subject_injury
NaN    9848
Yes    1631
No     1446
Name: count, dtype: int64

 subject_injury_cleaned
Missing    9848
Yes        1631
No         1446
Name: count, dtype: int64
Percentage of missing values:  76.19342359767892 %
Percentage of non-missing values:  23.806576402321085 %


force_type,Baton,Bodily Force,Chemical Irritant,Firearm,Gun Point Display,Improvised Weapon,Less Lethal,Less Lethal Projectile,Maximal Restraint Technique,Police K9 Bite,Taser
subject_injury_cleaned,,,,,,,,,,,
Missing,2,7051,1421,0,27,74,87,0,170,31,985
No,0,1093,131,2,33,34,0,1,0,2,150
Yes,2,1286,41,0,44,40,0,2,0,44,172


After calculation, about 76.2 percent of the subject_injury values are missing. This is concerning since most of the cases did not have a record. 
After cross-tabulating, we can see the missingness varies by the force type a lot. For example, the percentage of missingness in bodily force is high and has a lot of missing.

## *Phase 2: AI-assisted Revision
Keep Phase 1 frozen.

- **Phase 1 commit SHA ID:** [To be provided]
- **AI tool and model used:** [To be provided]

### *Prompts used

1. [Paste prompt]
2. [Paste follow-up prompt]

### *AI-proposed changes


[Paste the AI's concise numbered list verbatim.]

1. ...
2. ...

### *Evaluation of AI suggestions


| Change | Decision | Reason and verification |
|---|---|---|
| 1 | Accept / modify / reject | ... |
| 2 | Accept / modify / reject | ... |

### *AI-assisted solution in full
After the evaluation of AI suggestions, incorporate the accepted revisions into your Phase 1 solution and provide the final Phase 2 solution in full in the following cells.

In [ ]:
# Your Phase 2 solution to the problem goes here.


### *Reflection on AI assistance
In your own words without generative AI.

[In 150–300 words, describe the main improvements, AI mistakes or limitations, how the final solution was verified, and any remaining concerns.]